lib/README.md

```diff
@@ -0,0 +1,144 @@
+# Gutenberg PHP
+
+## File structure
+
+Gutenberg adds features to WordPress Core using PHP hooks and filters. Some
+features, once considered stable and useful, are merged into Core during a Core
+release. Some features remain in the plugin forever or are removed.
+
+To make it easier for contribtors to know which features need to be merged to
+Core and which features can be deleted, Gutenberg uses the following file
+structure for its PHP code:
+
+- `lib/experimental` - Experimental features that exist only in the plugin. They
+  are not ready to be merged to Core.
+- `lib/stable` - Stable features that exist only in the plugin. They could one
+  day be merged to Core, but not yet.
+- `lib/compat/wordpress-X.Y` - Stable features that are intended to be merged to
+  Core in the future X.Y release, or that were previously merged to Core in the
+  X.Y release and remain in the plugin for backwards compatibility when running
+  the plugin on older versions of WordPress.
+
+## Best practices
+
+### Prefer the `wp` prefix
+
+For features that may be merged to Core, it's best to use a `wp_` prefix for functions or a `WP_` prefix for classes. 
+
+This applies to both experimental and stable features.
+
+Using the `wp_` prefix facilitates the process of renaming functions and classes from `gutenberg_` to `wp_` if the feature is merged to Core.
+
+Functions that are intended solely for the plugin, e.g., plugin infrastructure, should use the `gutenberg_` prefix.
+
+#### Good
+
+```php
+function wp_get_navigation( $slug ) { ... }
+```
+
+#### Not so good
+
+```php
+function gutenberg_get_navigation( $slug ) { ... }
+```
+
+### Group PHP code by _feature_
+
+Developers should organize PHP into files or folders by _feature_, not by _component_.
+
+Also, developers should call `add_action` and `add_filter` immediately after the function being hooked is defined.
```

Comment:
Totally optional, but I found this sentence a little hard to parse at first. How about:

```suggestion
When defining a function that will be hooked, developers should call `add_action` and `add_filter` immediately after the function declaration.
```

Reply:
Sure!